vendor/bat-native-ledger/src/bat/ledger/internal/publisher/publisher_list_updater.cc

```diff
@@ -0,0 +1,159 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ledger/internal/publisher/publisher_list_updater.h"
+
+#include <memory>
+#include <utility>
+
+#include "bat/ledger/internal/common/time_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/publisher/publisher_list_reader.h"
+#include "bat/ledger/internal/request/request_publisher.h"
+#include "bat/ledger/internal/state/state_keys.h"
+#include "bat/ledger/option_keys.h"
+#include "net/http/http_status_code.h"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+using std::placeholders::_3;
+
+namespace {
+
+constexpr int64_t kRetryAfterFailureDelay = 150;
+constexpr int64_t kMaxRetryAfterFailureDelay = 4 * base::Time::kSecondsPerHour;
+
+}  // namespace
+
+namespace braveledger_publisher {
+
+PublisherListUpdater::PublisherListUpdater(bat_ledger::LedgerImpl* ledger)
+    : ledger_(ledger) {}
+
+PublisherListUpdater::~PublisherListUpdater() = default;
+
+void PublisherListUpdater::StartAutoUpdate() {
+  auto_update_ = true;
+  if (!timer_.IsRunning()) {
+    StartFetchTimer(FROM_HERE, GetAutoUpdateDelay());
+  }
+}
+
+void PublisherListUpdater::StopAutoUpdate() {
+  BLOG(1, "Cancelling publisher prefix list update");
+  auto_update_ = false;
+  timer_.Stop();
+}
+
+void PublisherListUpdater::SetOnPublisherListUpdatedCallback(
```

Comment:
could you tell a little bit more about this one?

Reply:
I modified it slightly so that `StartAutoUpdate` takes a callback instead of having this separate method, but the idea is that, ideally,`PublisherPrefixListUpdater` shouldn't have to know about the kinds of things that we might want to do after an update; it should just know how to perform the update, and let callers do what they want after each update (there are two callouts that happen in `publisher.cc`).